runtime/contract_update_validation.go

```diff
@@ -0,0 +1,434 @@
+		return validator.checkIdentifierEquality(foundType, expectedType)
+	}
+
+	// At this point, either both are qualified names, or both are simple names.
+	// Thus, do a one-to-one match.
+	if expectedType.Identifier.Identifier != foundType.Identifier.Identifier {
+		return false
+	}
+
+	return checkSliceEquality(expectedType.NestedIdentifiers, foundType.NestedIdentifiers)
+}
+
+func (validator *ContractUpdateValidator) checkIdentifierEquality(
+	qualifiedNominalType *ast.NominalType,
+	simpleNominalType *ast.NominalType) bool {
+
+	// Situation:
+	// qualifiedNominalType -> identifier: A, nestedIdentifiers: [foo, bar, ...]
+	// simpleNominalType -> identifier: foo,  nestedIdentifiers: [bar, ...]
+
+	// If the first identifier (i.e: 'A') refers to a composite decl that is not the enclosing contract,
+	// then it must be referring to an imported contract. That means the two types are no longer the same.
+	if qualifiedNominalType.Identifier.Identifier != validator.entryPointName {
+		return false
+	}
+
+	if qualifiedNominalType.NestedIdentifiers[0].Identifier != simpleNominalType.Identifier.Identifier {
+		return false
+	}
+
+	return checkSliceEquality(simpleNominalType.NestedIdentifiers, qualifiedNominalType.NestedIdentifiers[1:])
+}
+
+func (validator *ContractUpdateValidator) getTypeMismatchError(
+	expectedType ast.Type,
+	foundType ast.Type) error {
+
+	if validator.isInNestedDecl() {
+		return fmt.Errorf("type annotation does not match for field `%s` in `%s`. expected `%s`, found `%s`",
+			validator.currentFieldName,
+			validator.currentDeclName,
+			expectedType,
+			foundType)
+	}
+
+	return fmt.Errorf("type annotation does not match for field `%s`. expected `%s`, found `%s`",
+		validator.currentFieldName,
+		expectedType,
+		foundType)
+}
+
+// isInNestedDecl checks whether the currently visiting declaration is a nested one.
+func (validator *ContractUpdateValidator) isInNestedDecl() bool {
+	return validator.entryPointName != validator.currentDeclName
+}
+
+func getDeclarationName(declaration ast.Declaration) string {
+	switch decl := declaration.(type) {
+	case *ast.CompositeDeclaration:
+		return decl.Identifier.Identifier
+	case *ast.InterfaceDeclaration:
+		return decl.Identifier.Identifier
+	default:
+		panic(errors.NewUnreachableError())
+	}
+}
+
+func getDeclarationMembers(declaration ast.Declaration) *ast.Members {
+	switch decl := declaration.(type) {
+	case *ast.CompositeDeclaration:
+		return decl.Members
+	case *ast.InterfaceDeclaration:
+		return decl.Members
+	default:
+		panic(errors.NewUnreachableError())
+	}
+}
+
+func checkSliceEquality(expected []ast.Identifier, found []ast.Identifier) bool {
+	if len(expected) != len(found) {
+		return false
+	}
+
+	for index, element := range found {
+		if expected[index] != element {
+			return false
+		}
+	}
+	return true
+}
+
+func isQualifiedName(nominalType *ast.NominalType) bool {
+	return len(nominalType.NestedIdentifiers) > 0
+}
```

Comment:
Maybe just add this as a function to `ast.NominalType` itself

Reply:
added